// @useoptic/analytics/lib/events/tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExitedTaskWithLocalCli = exports.StartedTaskWithLocalCli = void 0;
const tslib_1 = require("tslib");
const RegisterEvent_1 = require("../interfaces/RegisterEvent");
const Events_1 = require("../interfaces/Events");
// @ts-ignore
const Joi = tslib_1.__importStar(require("@hapi/joi"));
require("joi-extract-type");
// Sent whenever an API is created
const StartedTaskWithLocalCliSchema = Joi.object({
    captureId: Joi.string().required(),
    cwd: Joi.string().required(),
    inputs: Joi.object().unknown(true),
});
exports.StartedTaskWithLocalCli = RegisterEvent_1.DescribeEvent(Events_1.Events.StartedTaskWithLocalCli, StartedTaskWithLocalCliSchema, (props) => `User ran task`);
const ExitedTaskWithLocalCliSchema = Joi.object({
    captureId: Joi.string().required(),
    interactionCount: Joi.number().required(),
    inputs: Joi.object().unknown(true),
});
exports.ExitedTaskWithLocalCli = RegisterEvent_1.DescribeEvent(Events_1.Events.ExitedTaskWithLocalCli, ExitedTaskWithLocalCliSchema, (props) => `User ran task`);
