// @useoptic/cli-shared/build/captures/avro/file-system/replicated-interactions-capture-saver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
//@ts-ignore
const oboe_1 = __importDefault(require("oboe"));
const capture_saver_1 = require("./capture-saver");
async function main(inputFilePaths, outputBaseDirectory, captureId, replicaCount) {
    console.log({ inputFilePaths });
    const events = await fs_extra_1.default.readJson(inputFilePaths.events);
    const captureBaseDirectory = path_1.default.join(outputBaseDirectory, '.optic', 'captures');
    const captureSaver = new capture_saver_1.CaptureSaver({
        captureBaseDirectory,
        captureId,
    });
    const input = fs_extra_1.default.createReadStream(inputFilePaths.interactions);
    await captureSaver.init();
    await new Promise((resolve, reject) => {
        oboe_1.default(input)
            .on('node', {
            // @ts-ignore
            '!.*': function (sample) {
                console.count('sample');
                console.log({ sample });
                Array.from({ length: replicaCount }).forEach(() => {
                    captureSaver.save(sample);
                });
            },
        })
            .on('done', function () {
            console.log('done');
            resolve();
        })
            .on('fail', function (e) {
            console.error(e);
            reject(e);
        });
    });
    const files = [
        {
            location: path_1.default.join(outputBaseDirectory, 'optic.yml'),
            contents: `name: ${JSON.stringify(path_1.default.basename(inputFilePaths.events))}`,
        },
        {
            location: path_1.default.join(outputBaseDirectory, '.optic', 'api', 'specification.json'),
            contents: JSON.stringify(events),
        },
        {
            location: path_1.default.join(outputBaseDirectory, '.optic', 'captures', captureId, 'optic-capture-state.json'),
            contents: JSON.stringify({
                captureId,
                status: 'completed',
                metadata: {
                    startedAt: new Date().toISOString(),
                    taskConfig: null,
                    lastInteraction: null,
                },
            }),
        },
    ];
    await Promise.all(files.map(async (x) => {
        const { location, contents } = x;
        await fs_extra_1.default.ensureDir(path_1.default.dirname(location));
        return fs_extra_1.default.writeFile(location, contents);
    }));
}
const [, , inputEventsFilePath, inputInteractionsFilePath, outputBaseDirectory, captureId, replicaCount,] = process.argv;
main({ interactions: inputInteractionsFilePath, events: inputEventsFilePath }, outputBaseDirectory, captureId, parseInt(replicaCount, 10));
