// @useoptic/cli-config/build/helpers/initial-task.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultCommandInit = exports.buildTask = void 0;
//@ts-ignore
const jsesc_1 = __importDefault(require("jsesc"));
function buildTask(apiName, flags, taskName) {
    const config = `
name: ${escapeIt(apiName)}
tasks:
${buildInitialTask(flags, taskName)}`.trimLeft();
    return config;
}
exports.buildTask = buildTask;
exports.defaultCommandInit = 'echo "Setup A Valid Command to Start your API!"';
function buildInitialTask(flags, taskName) {
    //default config and valid for start injected
    let commandConfig = `  ${taskName}:
     command: ${escapeIt(flags.command || exports.defaultCommandInit)}
     inboundUrl: ${flags.inboundUrl || 'http://localhost:4000'}
`.trimRight();
    if (flags.inboundUrl && flags.targetUrl) {
        commandConfig = `  ${taskName}:
     inboundUrl: ${flags.inboundUrl}
     targetUrl: ${flags.targetUrl}
`.trimRight();
    }
    return commandConfig;
}
function escapeIt(value) {
    const escaped = jsesc_1.default(value, { quotes: 'double' });
    if (escaped !== value) {
        return `"${escaped}"`;
    }
    return `"${value}"`;
}
