// @useoptic/analytics/lib/interfaces/Events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Events = void 0;
const tslib_1 = require("tslib");
//@ts-ignore
const keymirror_1 = tslib_1.__importDefault(require("keymirror"));
exports.Events = keymirror_1.default({
    //on-boarding
    UserLoggedInFromCLI: null,
    ApiCreated: null,
    ApiInitializedInProject: null,
    ApiCheckCompleted: null,
    //Errors
    JavascriptErrorDetectedInFrontend: null,
    RequirementForDiffsToHaveASuggestionFailed: null,
    //running tasks
    StartedTaskWithLocalCli: null,
    ExitedTaskWithLocalCli: null,
    //diffs
    NewBodyDiffRendered: null,
    BodyDiffRendered: null,
    UserChangedCaptureOverviewTab: null,
    UserEnabledInferPolymorphism: null,
    UserBeganAddingNewUrl: null,
    UserFinishedAddingNewUrl: null,
    UserResetDiff: null,
    ShowInitialDocumentingView: null,
    UpdateContribution: null,
    AddUrlModalNaming: null,
    AddUrlModalIdentifyingPathComponents: null,
    ShowCommitCard: null,
    UserPreviewedSuggestion: null,
    UserAcceptedSuggestion: null,
    SuggestionDisplayed: null,
    UserCommittedChanges: null,
});
