// @useoptic/cli-client/build/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const client_utilities_1 = require("@useoptic/client-utilities");
class Client {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
    }
    getIdentity() {
        const url = `${this.baseUrl}/identity`;
        return client_utilities_1.JsonHttpClient.getJsonWithoutHandlingResponse(url);
    }
    setIdentity(user) {
        if (user) {
            const url = `${this.baseUrl}/identity`;
            return client_utilities_1.JsonHttpClient.putJson(url, { user });
        }
    }
    postTrackingEvents(events) {
        const url = `${this.baseUrl}/tracking/events`;
        return client_utilities_1.JsonHttpClient.postJson(url, { events });
    }
    findSession(path, taskConfig, captureId) {
        const url = `${this.baseUrl}/sessions`;
        return client_utilities_1.JsonHttpClient.postJson(url, { path, taskConfig, captureId });
    }
    markCaptureAsCompleted(specId, captureId) {
        const url = `${this.baseUrl}/specs/${specId}/captures/${captureId}/status`;
        return client_utilities_1.JsonHttpClient.putJson(url, { status: 'completed' });
    }
    stopDaemon() {
        const url = `${this.baseUrl}/commands`;
        return client_utilities_1.JsonHttpClient.postJson(url, { type: 'shutdown' });
    }
}
exports.Client = Client;
